File: algorithms/2_medium/C/02044_count_number_of_maximum_bitwise_or_subsets/backtrack.py
class Solution:
    def subsets_backtrack(self, nums, curr_or, start_index):
        if curr_or == self.max_or:
            self.count += 2 ** (len(nums) - start_index);
            return

        for i in range(start_index, len(nums)):
            self.subsets_backtrack(nums, curr_or | nums[i], i + 1)

    def countMaxOrSubsets(self, nums: List[int]) -> int:
        self.max_or = 0
        for num in nums:
            self.max_or |= num
        self.count = 0
        self.subsets_backtrack(nums, 0, 0)

        return self.count

File: algorithms/2_medium/C/02044_count_number_of_maximum_bitwise_or_subsets/backtrack.rs
impl Solution {
    fn subsets_backtrack(nums: &Vec<i32>, max_or: i32, curr_or: i32, start_index: usize, count: &mut i32) {
        if curr_or == max_or {
            *count += 2_i32.pow((nums.len() - start_index) as u32);
            return;
        }
        for i in start_index..nums.len() {
            Self::subsets_backtrack(&nums, max_or, curr_or | nums[i], i + 1, count);
        }
    }
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let max_or = nums.iter().fold(0, |or, num| or | num);
        let mut count = 0;
        Self::subsets_backtrack(&nums, max_or, 0, 0, &mut count);

        count
    }
}
